cmd/holo/internal/generators.go

```diff
@@ -0,0 +1,173 @@
+	inputDir := getGenertorsDir()
+	if _, err := os.Stat(inputDir); err != nil {
+		if os.IsNotExist(err) {
+			return nil
+		}
+		return err
+	}
+	targetDir, err := getGeneratorCacheDir()
+	if err != nil {
+		return fmt.Errorf(
+			"couldn't access cache-dir ('%s') for generators: %s",
+			targetDir, err,
+		)
+	}
+	fmt.Fprintln(Stdout, "Processing generators...")
+	filepath.Walk(inputDir,
+		func(path string, info os.FileInfo, err error) error {
+			if err != nil {
+				Warnf(Stderr, "%s: %s", path, err.Error())
+				return nil
+			}
+			if isExecutableFile(info) {
+				fmt.Fprintln(Stdout, path)
+				if err = runGenerator(path, targetDir); err != nil {
+					Errorf(
+						Stderr,
+						"Running generator failed: %s", err.Error(),
+					)
+				}
+			}
+			return nil
+		})
+	for _, plugin := range config.Plugins {
+		if err := updatePluginPaths(plugin, targetDir); err != nil {
+			Errorf(Stderr,
+				"Failed to perpare generated dir for plugin '%s': %s",
+				plugin.id, err.Error(),
+			)
+		}
+	}
+	return nil
+}
+
+func updatePluginPaths(plugin *Plugin, dir string) error {
+	pluginDir := plugin.ResourceDirectory()
+	newPluginDir := filepath.Join(dir, plugin.id)
+	if info, err := os.Stat(newPluginDir); err == nil && info.IsDir() {
+		// Files were generated for this plugin.
+		// Fill the plugins directory with existsing static files.
+		if err := symlinkFiles(pluginDir, newPluginDir); err != nil {
+			return err
+		}
+		// Change the plugin resource dir to point to the generated dir.
+		resource, _ := filepath.Rel(RootDirectory(), dir)
+		plugin.SetResourceRoot(resource)
+	}
+	return nil
+}
+
+func symlinkFiles(oldDir string, newDir string) error {
+	return filepath.Walk(oldDir,
+		func(oldFile string, info os.FileInfo, err error) error {
+			if err != nil || oldFile == oldDir {
+				return err
+			}
+			relPath, _ := filepath.Rel(oldDir, oldFile)
+			newFile := filepath.Join(newDir, relPath)
+			err = os.Symlink(oldFile, newFile)
+			if os.IsExist(err) {
+				// newFile already exists. Examine it.
+				newFileInfo, err := os.Lstat(newFile)
+				if err == nil && info.IsDir() && !newFileInfo.IsDir() {
+					// newFile exists but is not a directory.
+					// If oldFile is a directory trying to symlink its contents
+					// will result in errors. Skip it.
+					return filepath.SkipDir
+				}
+				return nil
+			}
+			if err != nil {
+				return err
+			}
+			if info.IsDir() {
+				// Symlink to dir was created don't check its contents.
+				return filepath.SkipDir
+			}
+			return nil
+		})
+}
+
+func runGenerator(fileToRun string, targetDir string) error {
+	cmd := exec.Command(fileToRun)
+	env := os.Environ()
+	env = append(
+		env,
+		fmt.Sprintf("OUT=%s", targetDir),
+	)
+	cmd.Env = env
+	return cmd.Run()
+}
+
+func getGenertorsDir() string {
+	return filepath.Join(RootDirectory(), "/usr/share/holo/generators")
+}
+
+func getGeneratorCacheDir() (string, error) {
+	path, err := prepareDir(RootDirectory(), "/var/tmp/holo/generated")
```

Comment:
Since generated resource files are extremely transient, `/tmp` looks to be more appropriate than `/var/tmp`. I'd prefer having all generated files inside `$HOLO_CACHE_DIR` (like in the fallback path of this method) to simplify cleanup and to ensure that we don't accidentally reuse old files from a previous run.

Reply:
I wanted generated files to be cacheable (let the generator decide whether to regenerate or not) and avoid excessive regeneration. But thinking about it this leads to a problem when a generator just stops to generate a certain file it previously generated. Would you say we leave it to the generator to maintain his own cache and just always clean the generated files?